### PR TITLE
fix: #1408 Native Alert Patch for Webview

### PR DESCRIPTION
**Pull request by MarioAslau**
## **Description**

Currently on iOS when the user is promted with an alert to grant permissions for a website that has a big domain, it will hide the entire Alert message that the website is requesting.

## **Related issues**

https://github.com/MetaMask/mobile-planning/issues/1408

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **e2e Testing**
https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/abb68302-f423-4084-9cd5-3fef3a0a5c7c

## **Screenshots/Recordings**

Before:

This is the website we can use for debugging purposes: https://nfttoks.crypto.comsssssssss.go.f-secure.com.egghunter.in/camera.html

After:
Screenshot 2024-02-02 at 16 10 13

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ios/MetaMask.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1625,11 +1625,7 @@
 				IPHONEOS_DEPLOYMENT_TARGET = 11.0;
 				MTL_ENABLE_DEBUG_INFO = YES;
 				ONLY_ACTIVE_ARCH = YES;
-				OTHER_LDFLAGS = (
```
Comment (Cal-L) [inline ios/MetaMask.xcodeproj/project.pbxproj:1628]: Remove unrelated changes

File: ios/MetaMask.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -1671,11 +1667,7 @@
 				GCC_WARN_UNUSED_VARIABLE = YES;
 				IPHONEOS_DEPLOYMENT_TARGET = 11.0;
 				MTL_ENABLE_DEBUG_INFO = NO;
-				OTHER_LDFLAGS = (
```
Comment (Cal-L) [inline ios/MetaMask.xcodeproj/project.pbxproj:1674]: Remove unrelated changes

File: patches/react-native-webview+11.13.0.patch [hunk 3]
```diff
@@ -1225,7 +1225,59 @@ index 28c078a..6f7d0b7 100644
  #else
    NSAlert *alert = [[NSAlert alloc] init];
    [alert setMessageText:message];
-@@ -894,44 +900,49 @@ static NSDictionary* customCertificatesForHost;
+@@ -868,6 +874,51 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+ /**
+  * confirm
+  */
++
++- (void)webView:(WKWebView *)webView requestMediaCapturePermissionForOrigin:(WKSecurityOrigin *)origin initiatedByFrame:(WKFrameInfo *)frame type:(WKMediaCaptureType)type decisionHandler:(void (^)(WKPermissionDecision decision))decisionHandler  API_AVAILABLE(ios(15.0)){
```
Comment (Cal-L) [inline patches/react-native-webview+11.13.0.patch:1233]: Let's comment the reason for this change

File: patches/react-native-webview+11.13.0.patch [hunk 4]
```diff
@@ -1171,42 +1171,42 @@ index 0000000..b9581ac
...
-@@ -417,6 +419,7 @@ static NSDictionary* customCertificatesForHost;
+@@ -417,6 +419,7 @@ -(void)keyboardDisplacementFix
  - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context{
      if ([keyPath isEqual:@"estimatedProgress"] && object == self.webView) {
          if(_onLoadingProgress){
 +            _disablePromptDuringLoading = YES;
               NSMutableDictionary<NSString *, id> *event = [self baseEvent];
              [event addEntriesFromDictionary:@{@"progress":[NSNumber numberWithDouble:self.webView.estimatedProgress]}];
              _onLoadingProgress(event);
-@@ -492,6 +495,7 @@ static NSDictionary* customCertificatesForHost;
```
Comment (Cal-L) [inline patches/react-native-webview+11.13.0.patch:1201]: Are these just a byproduct of the patch or did customCertificatesForHost actually change?

**Review by Cal-L — COMMENTED** [review]
Left some comments

File: patches/react-native-webview+11.13.0.patch (hunk 3 above)
Comment (Cal-L) [inline patches/react-native-webview+11.13.0.patch:1233]: What happens on iOS<15?

File: patches/react-native-webview+11.13.0.patch [hunk 5]
```diff
@@ -1225,7 +1225,59 @@ index 28c078a..6f7d0b7 100644
...
++          deviceType = @"microphone";
++          break;
++      case WKMediaCaptureTypeCameraAndMicrophone:
++          deviceType = @"camera and microphone";
++          break;
++      default:
++          deviceType = @"unknown device";
++  }
++
++  NSString *message = [NSString stringWithFormat:@"The webpage %@ is requesting access to your %@. Do you want to allow this?", origin.host, deviceType];
```
Comment (Cal-L) [inline patches/react-native-webview+11.13.0.patch:1251]: Is there a way to access the native message? Is it possible to remove the title as to mimic the format of the original prompt before our changes?

Comment (jpcloureiro) [reply patches/react-native-webview+11.13.0.patch:1251]: Can we pass on this message through the JavaScript bridge?
Otherwise, we can't translate the string.

Comment (MarioAslau) [reply patches/react-native-webview+11.13.0.patch:1251]: @NicholasEllul this works the same as the original Alert, it will keep wrapping around the text so users can see it all

File: patches/react-native-webview+11.13.0.patch (hunk 4 above)
Comment (jpcloureiro) [reply patches/react-native-webview+11.13.0.patch:1201]: That is correct @Cal-L, these are just a byproduct of the patch.

The code after `@@ -492,6 +495,7 @@` is simply the hunk's header Git generates automatically

`customCertificatesForHost` will remain unchanged

File: patches/react-native-webview+11.13.0.patch [hunk 6]
```diff
@@ -1225,7 +1225,59 @@ index 28c078a..6f7d0b7 100644
...
++          deviceType = @"camera";
++          break;
++      case WKMediaCaptureTypeMicrophone:
++          deviceType = @"microphone";
++          break;
++      case WKMediaCaptureTypeCameraAndMicrophone:
++          deviceType = @"camera and microphone";
++          break;
++      default:
++          deviceType = @"unknown device";
```
[the remark this answers is not in the document]
Comment (MarioAslau) [reply patches/react-native-webview+11.13.0.patch:1249]: @NicholasEllul  besides that do you feel the current message displayed is enough from a security pov ?

File: patches/react-native-webview+11.13.0.patch (hunk 5 above)
Comment (MarioAslau) [reply patches/react-native-webview+11.13.0.patch:1252]: @NicholasEllul do you have a  longer valid website that I can possibly test besides our current example?

File: patches/react-native-webview+11.13.0.patch (hunk 6 above)
[the remark this answers is not in the document]
Comment (MarioAslau) [reply patches/react-native-webview+11.13.0.patch:1249]: @tommasini  
Screenshot 2024-02-21 at 10 43 55
Works well!

[the remark this answers is not in the document]
Comment (MarioAslau) [reply patches/react-native-webview+11.13.0.patch:1249]: @NicholasEllul we are not truncating the domain, we are using the regular iOS alert again. I don't suspect there will be a need for that as iOS handles it. WebView was imposing some limitations which was resulting in the bug

File: patches/react-native-webview+11.13.0.patch (hunk 3 above)
Comment (MarioAslau) [reply patches/react-native-webview+11.13.0.patch:1233]: @Cal-L @NicholasEllul this is what happens on iOS 14
Screenshot 2024-02-21 at 16 28 52

**Review by Cal-L — APPROVED** [review]
LGTM

File: patches/react-native-webview+11.13.0.patch (hunk 5 above)
Comment (MarioAslau) [reply patches/react-native-webview+11.13.0.patch:1251]: @NicholasEllul  seems to be able to handle it !
Screenshot 2024-02-26 at 13 59 27

**Review by NEllusion — APPROVED** [review]
Changes satisfy requirements from our team. Great work!
